src/app/server/AppDelegate.h

```diff
@@ -29,11 +29,31 @@ class AppDelegate
 public:
     virtual ~AppDelegate() {}
     /**
-     * This is called on start of session establishment process
+     * This is called when the PBKDFParamRequest is received and indicates the start of the session establishment process
      */
     virtual void OnCommissioningSessionEstablishmentStarted() {}
+
+    /**
+     * This is called when the commissioning session has been established
+     */
     virtual void OnCommissioningSessionStarted() {}
-    virtual void OnCommissioningSessionStopped(CHIP_ERROR err) {}
+
+    /**
+     * This is called when there is an error in establishing a commissioning session (such as, when an invalid passcode is provided)
+     *
+     * @param err CHIP_ERROR indicating the error that occurred during session establishment
+     *
+     * @return true if the commissioning window should be closed, false otherwise
+     */
+    virtual bool OnCommissioningSessionEstablishmentError(CHIP_ERROR err) { return false; }
+
+    /**
+     * This is called in addition to OnCommissioningSessionEstablishmentError i.e. when there is an error in establishing a
+     * commissioning session AND the commissioning window is closed. The window may be closed because
+     * OnCommissioningSessionEstablishmentError returned 'true', the commissioning attempts limit was reached or listening for PASE
+     * failed.
+     */
+    virtual void OnCommissioningSessionStopped() {}
```

Comment:
This should be called as well when the commissioning session is stopped due to no longer being used, so we have start/stop symmetry.

Reply:
OnCommissioningSessionStopped was a pre-existing API. Renaming it will cause build errors on existing consumers. Should I rename it anyway?

Reply:
I believe this comment is no longer applicable after the revert following this thread: https://github.com/project-chip/connectedhomeip/pull/28240#discussion_r1275398506